LIS2DH: Fixing few things related to multi-instance

avisconti opened the pull request:
This PR is completely fixing the IRQ handling for lis2dh multi-instance, as well as a couple of leftovers.

avisconti committed:
drivers/sensor: lis2dh: Fix interrupt handling for multi instance

Handle interrupts correctly for multi instance, checking per
device DT bindings instead of global Kconfig definitions.

The old interrupt behaviour has been maintained, fixing DRDY to
be on INT1 and Any Motion event on INT2.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

mbolivar-nordic commented on drivers/sensor/lis2dh/lis2dh_trigger.c:
```diff
@@ -395,47 +407,47 @@ int lis2dh_init_interrupt(const struct device *dev)
-		LOG_ERR("Cannot get pointer to %s device",
-			    LIS2DH_INT2_GPIO_DEV_NAME);
-		return -EINVAL;
+		LOG_INF("Cannot get pointer to irq2_dev_name");
```
Same comments here as for irq1_dev_name.

mbolivar-nordic commented on drivers/sensor/lis2dh/lis2dh_trigger.c:
```diff
@@ -444,13 +456,12 @@ int lis2dh_init_interrupt(const struct device *dev)
 	}
 
-	LOG_INF("int2 on pin=%d", LIS2DH_INT2_GPIOS_PIN);
+	LOG_INF("int2 on %s%d", cfg->irq2_dev_name, cfg->irq2_pin);
```
Same here as for irq1_dev_name

mbolivar-nordic commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -367,6 +368,12 @@ int lis2dh_init(const struct device *dev)
 			    &lis2dh_driver_api);
 
+#define IS_LSM303AGR_DEV(inst) \
+	DT_NODE_HAS_STATUS(DT_INST(inst, st_lsm303agr_accel), okay)
```
Did you mean to use `DT_NODE_HAS_COMPAT` instead?

avisconti replied:
Well, actually I'm not sure which is the best way. Here I need to understand whether a given instance for this driver is a LSM303AGR_ACCEL device or one of the others supported by the same driver. That's because there are some slight differences that should be handled.

If my undesratnd is correct DT_NODE_HAS_COMPAT is returning true if we have a given string in the compatibility, right? If so it seems reasonable to use this macro because in case of 303agr accel I would have:
    compatible = "st,lis2dh", "st,lsm303agr-accel";

Is that correct?

avisconti replied:
This point is not clear at all to me. Maybe I shouild take a look better to documentation

avisconti replied:
WHat about:
DT_NODE_HAS_COMPAT(DT_DRV_INST(inst), st_lsm303agr_accel)

It works fine on my side, and it looks more clear than before.

mbolivar-nordic replied:
>  DT_NODE_HAS_COMPAT(DT_DRV_INST(inst), st_lsm303agr_accel)

Yep, that's exactly what I had in mind, thanks

mbolivar-nordic commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -367,6 +368,12 @@ int lis2dh_init(const struct device *dev)
+	DT_NODE_HAS_STATUS(DT_INST(inst, st_lsm303agr_accel), okay)
+
+#define DISC_PULL_UP(inst) \
+	IS_ENABLED(DT_INST_PROP(inst, disconnect_sdo_sa0_pull_up))
```
Just use DT_INST_PROP; there's no need for IS_ENABLED. A boolean property is 0 or 1 in C depending on if the property is absent or present in DTS.

See the DT_PROP docs: https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_PROP

avisconti replied:
OK, got it.
Thanks

avisconti commented on drivers/sensor/lis2dh/lis2dh_trigger.c:
```diff
@@ -367,17 +376,20 @@ static void lis2dh_work_cb(struct k_work *work)
-		LOG_ERR("Cannot get pointer to %s device",
-			    LIS2DH_INT1_GPIO_DEV_NAME);
-		return -EINVAL;
+		LOG_INF("Cannot get pointer to irq1_dev_name");
```
Well, actually, since we may have instances that defines irq in DT and other that don't, I'm no longer treating this as an error. Rather, I'm returnin ok (so LOG_INF is also used instead of LOG_ERR) and lis2dh->gpio_int1 is used in other calls as a matter to discriminate instances that are using interrupts.

Moreover cfg->irq1_dev_name is NULL, so I cannot use it.

mbolivar-nordic replied:
> Moreover cfg->irq1_dev_name is NULL, so I cannot use it.

Not always, right? If the GPIO driver is not compiled in, the name pointer could be non-null and the device could still not exist.

mbolivar-nordic replied:
If that's not a concern for you, OK.

avisconti replied:
Yes, I understand there might be false positives. Nevertheless I think we might accept this.

avisconti commented on drivers/sensor/lis2dh/lis2dh_trigger.c:
```diff
@@ -367,17 +376,20 @@ static void lis2dh_work_cb(struct k_work *work)
-			    LIS2DH_INT1_GPIO_DEV_NAME);
-		return -EINVAL;
+		LOG_INF("Cannot get pointer to irq1_dev_name");
+		status = 0;
```
If I'm falling in this case I'm assuming that the irq part is not present in the DT for that particular instance. In the dts bindings I defined it as 'not required'

mbolivar-nordic replied:
> If I'm falling in this case I'm assuming that the irq part is not present in the DT for that particular instance. In the dts bindings I defined it as 'not required'

OK, thanks for explaining

avisconti commented on drivers/sensor/lis2dh/lis2dh.h:
```diff
@@ -207,6 +192,14 @@ struct lis2dh_config {
 	const union lis2dh_bus_cfg bus_cfg;
+#ifdef CONFIG_LIS2DH_TRIGGER
+	const char *irq1_dev_name;
+	uint8_t irq1_pin;
```
Yes, sure. I'll change it! Thanks.

avisconti commented on drivers/sensor/lis2dh/lis2dh.h:
```diff
@@ -207,6 +192,14 @@ struct lis2dh_config {
+#ifdef CONFIG_LIS2DH_TRIGGER
+	const char *irq1_dev_name;
+	uint8_t irq1_pin;
+	uint8_t irq1_flags;
```
Ack. I'll change it as well.

avisconti commented on drivers/sensor/lis2dh/lis2dh.h:
```diff
@@ -207,6 +192,14 @@ struct lis2dh_config {
+	uint8_t irq1_pin;
+	uint8_t irq1_flags;
+	const char *irq2_dev_name;
+	uint8_t irq2_pin;
+	uint8_t irq2_flags;
```
ack

avisconti commented on drivers/sensor/lis2dh/lis2dh_trigger.c:
```diff
@@ -395,47 +407,47 @@ int lis2dh_init_interrupt(const struct device *dev)
 	}
 
-	LOG_INF("int1 on pin=%d", LIS2DH_INT1_GPIOS_PIN);
+	LOG_INF("int1 on %s%d", cfg->irq1_dev_name, cfg->irq1_pin);
```
Mmmmh. Looks like GPIOA0, GPIOC0, GPIOC1. Something like that. But that because I'm using the shield on top of a nucleo board. Maybe I can do something like "GPIOA(0)" or "GPIOA (0)". What do you think about it?

pabigot replied:
I personally like `%s.%02u` for peripheral+pin formatting.

avisconti replied:
I completely forgot about this point. I will follow the suggestion of @pabigot, which seems a nice and clear way to represent this kind of resources.

avisconti commented:
@mbolivar-nordic 
Thanks for reviewing. I implemented few of all requests.
Other seems to be not approriate (see comments)
And still there is one of your comments open which I need to go deeper through it (DT_NODE_HAS_COMPAT)

avisconti committed:
drivers/sensor: lis2dh: Fix multi-instance leftovers

Introduced two per instance booleans, which get set using DT info:

  1. is_lsm303agr_dev: if the device is a LSM303AGR_ACCEL, then
     the scale values have to be changed.

  2. Handle disconnect-sdo-sa0-pull-up if present in ntsnace DT.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
dts/bindings: Fix lis2dh12 and lis3dh

LIS3DH and LIS2DH12 devices share the same driver with
LIS2DH device. So, the bindings are reusing the same
common structure (as LSM30AGR-ACCEL and LSM303DLHC-ACCEL
devices are already doing)

Signed-off-by: Armando Visconti <armando.visconti@st.com>

mbolivar-nordic commented:
@avisconti thanks for the update; I've resolved all but 2 comments

avisconti commented:
> @avisconti thanks for the update; I've resolved all but 2 comments

@mbolivar-nordic 
Thanks for your review. I appreciate it!

pabigot reviewed (approved):
Looks good.

If you have to touch this again, please fix the commit message for `drivers/sensor: lis2dh: Fix multi-instance leftovers` as the word `ntsnace` is difficult to recognize.

WilliamGFish commented:
@avisconti : I have raised a change request for this driver.(sensor: driver: lis2dh interrupt definitions #30133)

It would good if the user was able to define which interrupt they wished to use for motion. For example the Thingy:52 using INT1 as standard, it could use INT2 but requires physical intervention, therefore the driver does not allow the user to define the trigger type for sensor. As far as I am aware the device itself will allow either or both interrupts to be drdy or motion.

Billy..
